test/test_trio_server/mock_lean.py

```diff
@@ -0,0 +1,175 @@
+
+
+@dataclass
+class LeanShouldGetRequest(LeanScriptStep):
+    """
+    Used to check that Lean gets the request it is looking for from Lean-Python interface.
+    This version is less specific.  It checks that the request is of the correct type and
+    has the desired fields.
+    """
+    request: Request
+    seq_num: int
+
+    async def run(self, server: 'MockLeanServerProcess') -> Awaitable[None]:
+        self.request.seq_num = self.seq_num
+        message = json.loads(self.request.to_json())
+        print(f"\nLean should receive the following request:\n{message}")
+        return await server.assert_message_is_received(message)
+
+
+@dataclass
+class LeanSendsResponse(LeanScriptStep):
+    """
+    Simulate Lean's output behaviour
+    """
+    message: Dict  # should be JSON encodable
+
+    async def run(self, server: 'MockLeanServerProcess') -> None:
+        print(f"\nLean sends the following response:\n{self.message}")
+        server.send_message(self.message)
+
+
+@dataclass
+class LeanSendsBytes(LeanScriptStep):
+    """
+    Simulate Lean's output behaviour at the byte level.
+    Doesn't include new line characters unless passed in.
+    """
+    message_bytes: bytes
+
+    async def run(self, server: 'MockLeanServerProcess') -> None:
+        print(f"\nLean sends the following bytes:\n{self.message_bytes!r}")
+        server.send_bytes(self.message_bytes)
+
+@dataclass
+class LeanShouldNotGetRequest(LeanScriptStep):
+    """
+    Check that Lean has not received any requests (perhaps sent prematurely).
+    """
+    async def run(self, server: 'MockLeanServerProcess') -> Awaitable[None]:
+        print(f"\nLean should not have received any requests yet.")
+        return await server.assert_no_messages_received()
+
+
+
+class MockLeanServerProcess(trio.Process):
+
+    def __init__(self, script: List[LeanScriptStep]):
+        self.stdin = trio.testing.MemorySendStream()      # a stream for mock lean to read from
+        self.stdout = trio.testing.MemoryReceiveStream()  # a stream for mock lean to write to
+        self.messages: Deque[Dict] = deque()
+        self.partial_message: bytes = b""
+        self.script: List[LeanScriptStep] = script
+
+    def kill(self):
+        pass
+
+    @staticmethod
+    def parse_message(b: bytes) -> Dict:
+        return json.loads(b.decode())
+
+    @staticmethod
+    def messages_are_equal(m1: Dict, m2: Dict):
+        s1 = json.dumps(m1, sort_keys=True)
+        s2 = json.dumps(m2, sort_keys=True)
+        return s1 == s2
+
+    def collect_stdin_messages(self):
+        try:
+            data = self.stdin.get_data_nowait()
+        except trio.WouldBlock:  # no data
+            return
+
+        raw_messages = (self.partial_message + data).split(b"\n")
+        self.partial_message = raw_messages.pop()
+        self.messages.extend(self.parse_message(m) for m in raw_messages)
+
+    async def assert_message_is_received(self, message_expected):
+        await trio.sleep(0.01)
```

Comment:
Why do we need this explicit sleep (and the one below)?

Reply:
That is a great question.  I'm pretty sure it was necessary, but let me remember/investigate why and I'll add it as a comment to the code.

Reply:
Ok, I've addressed this in this commit (https://github.com/leanprover-community/lean-client-python/pull/3/commits/a585707f59fd7fcabf7a183ec0b3268dbf5df0ee).  In particular, 
1. I moved the `sleep` command into the `collect_stdin_messages` to avoid duplicated code.
2. I clearly explained what three goals the `sleep` command (in conjunction with `get_data_nowait()`) are intended to accomplish.
3. I made the waiting time a parameter which could be changed.  It might need to be extended if we are testing some complicated functionality which takes a while to compute.
4. I left the default of 0.01 seconds.  In my tests, 0.001 seconds also allowed the tests to pass but 0.0001 was too fast.  I thought 0.01 seconds was short enough to not make the tests take too long, but also gave an order of magnitude buffer to account for slow machines.

Reply:
I still don't understand. The whole point of this async thing is to avoid guessing how long we should wait for answers from Lean. Why do we need any sleep? Is this only for tests, simulating some delay in answers, or is this going to be needed in actual uses of this lean-python client?

Reply:
In short, it is only needed for tests because we don't want to wait forever for the Lean Python client to return information which it may never do.

Remember, this is "Mock Lean", so I'm simulating a fake Lean processes waiting for our Python client to set information.  A real Lean client would make a blocking call to read the data and wait as long as necessary.  That could be accomplished via `self.stdin.get_data()` in `collect_stdin_messages`, but the problem is that I don't know if there are any messages and I could be left waiting forever.  This is used to support `LeanShouldGetRequest` and `LeanShouldNotGetRequest`.  Those test that either Lean receives a request (and it a particular form) or that it doesn't receive a request, so I need the ability to stop waiting.

So what I'm doing here is waiting for our Python client for a little bit (enough time for our Python client to run and send a message if it is going to), and then checking if the message is sent in a non-blocking manner.

Another way I could possibly implement this is via a blocking call, `self.stdin.get_data()`, and then use a feature of Trio testing framework that has some way to stop execution of a blocked coroutine after a certain time.  I think this might exist, however, I'd have to look into that more.  It is not clear to me how to implement this.

Reply:
So like I said, an alternative way to do this would be with [cancellation and timeouts](https://trio.readthedocs.io/en/stable/reference-core.html#cancellation-and-timeouts).  I would still need to look into how to do this more.  Would you prefer that approach?

Reply:
I could do something like this inside `collect_stdin_messages`:
```
with trio.move_on_after(wait_time_seconds):
    try:
        data = await self.stdin.get_data()
    except Cancelled:
        return None
   
    self.partial_message = raw_messages.pop()
    self.messages.extend(self.parse_message(m) for m in raw_messages)
```

Reply:
Ok, I just pushed [a commit using the above snippet](https://github.com/leanprover-community/lean-client-python/pull/3/commits/1c9caac4dd845324d2e2a208a8c69dd3f5121d87).  I think it will clear up a lot of the confusion.

Reply:
Yes, I think this expresses what's going on much more directly.